File: AlgorithmsPackPythonSolutions/Similar String Groups.py
class UnionFind:
    def __init__(self, n):
        self.forests = n
        self.rank = [1] * n
        self.parent = list(range(n))

    def link(self, x, y):
        if self.rank[x] > self.rank[y]:
            x, y = y, x
        self.parent[x] = y
        if self.rank[x] == self.rank[y]:
            self.rank[y] += 1

    def find_set(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find_set(self.parent[x])
        return self.parent[x]

    def union_set(self, x, y):
        x = self.find_set(x)
        y = self.find_set(y)
        if x != y:
            self.link(x, y)
            self.forests -= 1
        return x != y


class Solution:
    def isSimilar(self, a: str, b: str) -> bool:
        diff = 0
        for i in range(len(a)):
            if a[i] != b[i]:
                diff += 1
            if diff > 2:
                return False
        return True

    def numSimilarGroups(self, strs: List[str]) -> int:
        n = len(strs)
        uf = UnionFind(n)
        for i in range(n):
            for j in range(i + 1, n):
                if self.isSimilar(strs[i], strs[j]):
                    uf.union_set(i, j)
        return uf.forests

File: AlgorithmsPackCPP/Similar String Groups.cpp
class UnionFind{
    private:
    vector<int>rank,parent;

    void link(int x, int y){
        if(rank[x] > rank[y])
            swap(x,y);
        parent[x] = y;
        if(rank[x]==rank[y])
            rank[y]++;
    }

    public:
    int forests;

    UnionFind(int n){
        forests = n;
        rank = vector<int>(n);
        parent = vector<int>(n);

        for(int i = 0; i < n; i++){
            rank[i] = 1;
            parent[i] = i;
        }
    }

    int find_set(int x){
        if(parent[x]==x)
            return x;
        return parent[x] = find_set(parent[x]);
    }

    int union_set(int x, int y){
        x = find_set(x);
        y = find_set(y);
        if(x!=y){
            link(x,y);
            --forests;
        }
        return x!=y;
    }
};

class Solution {
public:
    bool isSimilar(string&a,string&b){
        int diff = 0;
        for(int i = 0; i < a.size(); i++){
            if(a[i]!=b[i])
                diff++;
            if(diff>2)
                return false;
        }
        return true;
    }

    int numSimilarGroups(vector<string>& strs) {
        int n = strs.size();
        UnionFind uf(n);
        for(int i = 0; i < n; i++){
            for(int j = i+1; j < n; j++){
                if(isSimilar(strs[i],strs[j]))
                    uf.union_set(i,j);
            }
        }
        return uf.forests;
    }
};
